Rickord, Jake JJ<br>
Project 3: Spotify Song Classifications<br>

Import needed packages

In [118]:
import os
import pandas as pd
import numpy as np
import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
from spotipy.oauth2 import SpotifyOAuth
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import precision_recall_fscore_support as score

Authenticate requests

In [119]:
scope = "user-library-read"

auth_manager = SpotifyClientCredentials(client_id='######', client_secret='#########')
sp = spotipy.Spotify(client_credentials_manager = auth_manager)

Extract track listings from liked playlist

In [120]:
track_uris = []
playlist_uri='3gfQ9CScIyMxMZwphYM4cZ'
pl_pull = sp.playlist_tracks(playlist_uri)['items']
for track in pl_pull:
    track_uris.append(track['track']['uri'])

In [121]:
print(track_uris)

['spotify:track:36lV4io8Gc69fDinNaazAg', 'spotify:track:0YOOXnCJihgyluizqhAcrz', 'spotify:track:49YQ9Mq8CKelQsmjdNiICu', 'spotify:track:1vXHRgrkWnCMab8dhVPh0f', 'spotify:track:3h5AZDf5z7D18plaLtHTfi', 'spotify:track:1tQGRq2WOBXjL3JWdWMONg', 'spotify:track:3UCnS94uOqgvTmHohbgH9W', 'spotify:track:1WSEWiuB7KeEchBhvRb7qX', 'spotify:track:4z5kF7eIQhc6cvdb2MxEDD', 'spotify:track:1kStecVGAvK2RLhf0S9ywc', 'spotify:track:5tN0LIuQDapSFedlYJuG0k', 'spotify:track:3R6yNicsZrWF8ybl02imcB', 'spotify:track:0YnbCBn2UdoM5Y1l6Am5EY', 'spotify:track:4a2uqVlpRChHj32EjJLu7G', 'spotify:track:7rwKNqgVO0OfdOLNSKfYR4', 'spotify:track:4Oz6xl8qhJTdjYZeT98V4m', 'spotify:track:48nZLOpJOINkICPWpUuhaN', 'spotify:track:6P7kZB7hSJFyaYJcu3cYQJ', 'spotify:track:0Ld3o5VR3A5c8Je92xFG4e', 'spotify:track:3BCsNM2w6HcCbrVWlhjTpk', 'spotify:track:3Ozx6IrGdoQyAworJzvBDE', 'spotify:track:3t5UYY2uXE6CIEd6FYXrho', 'spotify:track:5Os2lSlPvvc40Wmun45Tz8', 'spotify:track:62EcdBCqjB62CxnhgBkkJT', 'spotify:track:5XpJVr7FpYkZrgPCHbFF4E',

In [122]:
print(len(track_uris))

100


Looks like we only grabbed the first 100, let's try to grab the rest of the entries as well, looks like from documentation on spotipy we'll need to continue with other pages of return.

In [123]:
track_uris = []
playlist_uri='3gfQ9CScIyMxMZwphYM4cZ'
#grabs return
pl_pull = sp.playlist_tracks(playlist_uri)
#page 1 of return
track_list = pl_pull['items']
#grabs all pages, adds them to track_list
while pl_pull['next']:
    pl_pull = sp.next(pl_pull)
    track_list.extend(pl_pull['items'])
#just grab the uris from it
for track in track_list:
    track_uris.append(track['track']['uri'])

In [124]:
print(len(track_uris))

568


Now it looks like we've grabbed the full playlist. Let's grab their sound details now for each track, and add these details into a dataframe.

First let's set up the dataframe with the style we'd like it in

In [125]:
track_features = list(sp.audio_features(track_uris[0])[0].keys())
track_features.insert(0, 'track_uri')
track_features.append('Liked?')
df = pd.DataFrame(columns = track_features)
print(df)

Empty DataFrame
Columns: [track_uri, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, type, id, uri, track_href, analysis_url, duration_ms, time_signature, Liked?]
Index: []


Looks great, now let's add our songs to it

In [126]:
for track in track_uris:
    i_tf = list(sp.audio_features(track)[0].values())
    i_tf.insert(0, track[track.index('track:')+len('track:'):])
    i_tf.append(1)
    df.loc[len(df)] = i_tf
print(df)

                  track_uri  danceability  energy  key  loudness  mode  \
0    36lV4io8Gc69fDinNaazAg         0.591   0.411    9    -7.496     1   
1    0YOOXnCJihgyluizqhAcrz         0.495   0.434    2    -8.079     1   
2    49YQ9Mq8CKelQsmjdNiICu         0.470   0.220    3   -11.727     1   
3    1vXHRgrkWnCMab8dhVPh0f         0.490   0.180    7   -13.332     1   
4    3h5AZDf5z7D18plaLtHTfi         0.632   0.254    3   -10.898     1   
..                      ...           ...     ...  ...       ...   ...   
563  1UVaHJQCmYxBK4XpYnYazU         0.519   0.691    9   -10.254     0   
564  4ma2AbPdsy5ijCyJn16mAv         0.515   0.418    2   -11.794     1   
565  5i4cie6F2AD20KiZ3XxUTq         0.371   0.640    9    -9.154     0   
566  3e9dlCDa68YOOJ5kFSiEdp         0.524   0.733    0    -7.390     1   
567  77R5bi9hfAUKZmu6tjOcxg         0.470   0.133   11   -13.678     1   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0         0.0299        0.7130 

Looking spectacular. Now we have our dataset for songs we like. Now let's grab a dataset of songs we dislike. We've sampled some here in a collection: "6kT6nn9Vd1e52XRG7OS2IE"

In [127]:
bad_track_uris = []
bad_playlist_uri='6kT6nn9Vd1e52XRG7OS2IE'
bad_pl_pull = sp.playlist_tracks(bad_playlist_uri)['items']
for track in bad_pl_pull:
    bad_track_uris.append(track['track']['uri'])

And let's add these to that dataframe

In [128]:
for track in bad_track_uris:
    i_tf = list(sp.audio_features(track)[0].values())
    i_tf.insert(0, track[track.index('track:')+len('track:'):])
    i_tf.append(0)
    df.loc[len(df)] = i_tf
print(df)

                  track_uri  danceability  energy  key  loudness  mode  \
0    36lV4io8Gc69fDinNaazAg         0.591   0.411    9    -7.496     1   
1    0YOOXnCJihgyluizqhAcrz         0.495   0.434    2    -8.079     1   
2    49YQ9Mq8CKelQsmjdNiICu         0.470   0.220    3   -11.727     1   
3    1vXHRgrkWnCMab8dhVPh0f         0.490   0.180    7   -13.332     1   
4    3h5AZDf5z7D18plaLtHTfi         0.632   0.254    3   -10.898     1   
..                      ...           ...     ...  ...       ...   ...   
613  1LJYn86ysceH708AIkw0VZ         0.975   0.482    9    -7.940     1   
614  7FERnkDYqD9D9DJdNjQlnf         0.761   0.246    0   -11.676     1   
615  4F0flTfcEU2ZcStYhJsaRy         0.862   0.717    9    -7.158     1   
616  15hJmqqEtASVXl6sM7i4UF         0.615   0.600   10    -5.620     1   
617  0lTqQTHcYb0RzGa5STjVrf         0.821   0.714    2    -5.977     1   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0         0.0299       0.71300 

And now our dataset is complete with song info! Let's drop some of those auxilliary columns that aren't related to song information now.

In [129]:
df.columns

Index(['track_uri', 'danceability', 'energy', 'key', 'loudness', 'mode',
       'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'type', 'id', 'uri', 'track_href', 'analysis_url',
       'duration_ms', 'time_signature', 'Liked?'],
      dtype='object')

In [130]:
df = df.drop(['id', 'type', 'uri', 'track_href', 'analysis_url'], axis=1)

In [131]:
print(df)

                  track_uri  danceability  energy  key  loudness  mode  \
0    36lV4io8Gc69fDinNaazAg         0.591   0.411    9    -7.496     1   
1    0YOOXnCJihgyluizqhAcrz         0.495   0.434    2    -8.079     1   
2    49YQ9Mq8CKelQsmjdNiICu         0.470   0.220    3   -11.727     1   
3    1vXHRgrkWnCMab8dhVPh0f         0.490   0.180    7   -13.332     1   
4    3h5AZDf5z7D18plaLtHTfi         0.632   0.254    3   -10.898     1   
..                      ...           ...     ...  ...       ...   ...   
613  1LJYn86ysceH708AIkw0VZ         0.975   0.482    9    -7.940     1   
614  7FERnkDYqD9D9DJdNjQlnf         0.761   0.246    0   -11.676     1   
615  4F0flTfcEU2ZcStYhJsaRy         0.862   0.717    9    -7.158     1   
616  15hJmqqEtASVXl6sM7i4UF         0.615   0.600   10    -5.620     1   
617  0lTqQTHcYb0RzGa5STjVrf         0.821   0.714    2    -5.977     1   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0         0.0299       0.71300 

Okay, now with our dataset all rev'd up, time to split it into training and test sets!

In [169]:
X = df.drop(['track_uri', 'Liked?'], axis=1)
y = df['Liked?']
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)
print(X_train)
print(y_train)

     danceability  energy  key  loudness  mode  speechiness  acousticness  \
89          0.575   0.765    9    -6.608     1       0.0345       0.00703   
285         0.484   0.793    2    -6.797     1       0.2110       0.00069   
310         0.625   0.715    7    -4.539     0       0.0942       0.04130   
367         0.529   0.934    0    -4.808     1       0.0602       0.00153   
46          0.410   0.233    0   -12.878     1       0.0286       0.83100   
..            ...     ...  ...       ...   ...          ...           ...   
277         0.674   0.204    5   -13.535     1       0.0384       0.88700   
9           0.544   0.224    7   -14.410     1       0.0351       0.78100   
359         0.397   0.903    2    -4.577     1       0.0457       0.00575   
192         0.708   0.413    7   -10.856     1       0.1010       0.78300   
559         0.499   0.160    7   -13.632     1       0.0362       0.89900   

     instrumentalness  liveness  valence    tempo  duration_ms  time_signat

Alrighty, now with that being broken up, we shall train our model and evaluate

Let's try out Logistic Regression, which will likely make the most sense considering our output of Liked? is binary, either yes or no.

In [170]:
logmodel = LogisticRegression()
logmodel.fit(X_train,y_train)
predictions = logmodel.predict(X_test)

In [171]:
print(classification_report(y_test,predictions));

              precision    recall  f1-score   support

           0       0.00      0.00      0.00        13
           1       0.92      1.00      0.96       142

    accuracy                           0.92       155
   macro avg       0.46      0.50      0.48       155
weighted avg       0.84      0.92      0.88       155



Interesting, it predicted that I don't dislike any songs. My tastes are decently diverse, but I think this is more likely due to the severe imbalance between the size of the liked songs playlist (about 550) and the size of the disliked songs playlist (about 60). Let's see if we can't use another algorithm to take into account some of this imbalance. The first one that comes to mind if the SVC model.

In [172]:
from sklearn.svm import SVC

In [173]:
svcmodel = SVC(kernel='linear', class_weight = {0:0.92, 1:0.08})
svcmodel.fit(X_train, y_train)
predictions = svcmodel.predict(X_test)
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.08      0.62      0.14        13
           1       0.91      0.37      0.52       142

    accuracy                           0.39       155
   macro avg       0.50      0.49      0.33       155
weighted avg       0.84      0.39      0.49       155



Looks like even with weightings inverse to their proportions our f-1 and accuracy score aren't great. Let's test a few different weights out and see what the ideal weighting is and the resulting f-1 score.

In [174]:
weight0 = 0
weight1 = 1
bestweight0 = 0
besweight1 = 1
bestf1 = 0
for asweight in np.arange(0, 1, 0.05):
    svcmodel = SVC(kernel='linear', class_weight = {0:weight0+asweight, 1:weight1-asweight})
    svcmodel.fit(X_train, y_train)
    predictions = svcmodel.predict(X_test)
    precision,recall,fscore,support=score(y_test,predictions,average='macro')
    if fscore>bestf1:
        bestf1=fscore
        bestweight0 = weight0+asweight
        bestweight1 = weight1-asweight

print(bestf1)
print(bestweight0)
print(bestweight1)

0.4894598155467721
0.8500000000000001
0.1499999999999999


In [181]:
svcmodel = SVC(kernel = 'linear', class_weight = {0:0.85, 1:0.15})
svcmodel.fit(X_train, y_train)
predictions = svcmodel.predict(X_test)
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       0.08      0.15      0.11        13
           1       0.92      0.84      0.88       142

    accuracy                           0.78       155
   macro avg       0.50      0.50      0.49       155
weighted avg       0.85      0.78      0.81       155



In [153]:
import warnings
warnings.filterwarnings('ignore')

Looks like even with some balancing involved, we're still getting model outputs with best accuracy when disliked songs are simply excluded. Let's try taking a random sampling of our liked songs to try to even out the number of songs liked versus disliked.

In [175]:
bdf = pd.DataFrame(columns = df.columns)

Grab random 100 rows from liked playlist

In [176]:
bdf = df.loc[df['Liked?']==1].sample(n=50)
bdf

track_uri  danceability  energy  key  loudness  mode  \
104  6mPpHS9ocuKjsehu8NgD7r         0.430   0.501    9    -8.864     1   
521  4urpFW2JCm486dHJmsQ8Ps         0.607   0.336    3   -11.854     1   
526  7c6TRb4cvgHWVhTJDfRqth         0.598   0.898   10    -5.155     1   
38   5AeoHJUx0PJXAzN425xryh         0.413   0.100    0   -16.349     1   
469  0M3adYbGtyRHACP86dey1H         0.426   0.888    0    -3.720     0   
153  3VGjFO89Sh53UjTRGJXWmi         0.529   0.497    2    -8.979     1   
424  1Jj6MF0xDOMA3Ut2Z368Bx         0.724   0.436    0    -9.321     1   
102  2Zgnaip1c876zmBhz9HifI         0.544   0.745    3    -3.401     1   
248  2gA74HvN6NKFrhgzpd5oNE         0.645   0.486    0    -7.299     1   
191  5zXNvJgCNNqmC262XJp8TW         0.663   0.649    9    -4.808     1   
432  54ugrImfl3YTyfYEFFi4xq         0.467   0.966    7    -3.915     1   
66   1kAvkf8WQgVGuwxvSqrVwQ         0.654   0.425    0    -9.432     1   
552  06YV4yr9sdJqSNj4HjZk2s         0.230   0.406    6    -9.984     0   
493  6p2liQLGoDaLXgND68bfVt         0.687   0.677    2    -4.172     1   
193  1kqpApNdM0egGfSNBlm8Md         0.656   0.379    0   -10.529     1   
135  757S467W1kCc0ZUCRwXmA2         0.585   0.661    9    -5.053     1   
217  7F9zSVVD38CDt8ZMpe6fFg         0.722   0.438    2    -8.386     1   
383  70wYA8oYHoMzhRRkARoMhU         0.467   0.988   11    -3.313     1   
142  2AwPXDmMeUJGt2Ro5Ea10G         0.529   0.338    8    -8.608     1   
292  3wXHHCV4A7KsoTlHZWVZPd         0.393   0.558    3    -9.429     1   
128  7cpz6hJMwGp4WWupErggGk         0.524   0.388    7   -10.809     1   
399  6gxycjJNMgmAyfzUXBN80P         0.432   0.261    4   -12.591     1   
219  0Ujit1HUO8x7Hl79EIqRXz         0.698   0.517    7    -8.272     1   
483  21jGcNKet2qwijlDFuPiPb         0.695   0.762    0    -3.497     1   
228  1zZIVe9x0JhdhpuQjLTjIM         0.567   0.542    2    -6.640     1   
561  1KDPWUuA213r6M2WbWUauE         0.359   0.621   11    -7.482     0   
445  4InNoUpsnbiq97FLvir4lU         0.645   0.946    4    -2.486     1   
12   0YnbCBn2UdoM5Y1l6Am5EY         0.666   0.412    7   -14.082     1   
188  4jjvUXBBqQDFRGX4SR5OhS         0.690   0.657    7    -4.397     0   
291  3QKcv32c31JrHs1YqtnqRB         0.509   0.881    2    -3.169     1   
305  6RT2boG8dbLXpgnSXQx1LY         0.403   0.960    4    -2.708     0   
186  1Q9BAEptV3nrj8sXBHU70M         0.422   0.651    7    -6.867     1   
500  2F6FfZ4w8z3eJpSxPotVO5         0.468   0.769    4    -7.036     1   
453  11ozIUBoXAgRTVWelDn4pL         0.703   0.731    8    -6.283     0   
397  0rUNZQuYQvOz6A6zwyT6tM         0.242   0.983    2    -2.679     1   
339  28IEbk5a7twNTbUEvWslUb         0.644   0.939   11    -3.486     0   
462  1yjY7rpaAQvKwpdUliHx0d         0.602   0.923    5    -3.763     1   
382  71iSmEeF0qRVyULABxP75P         0.519   0.799   10    -4.207     1   
226  5N0G5cRxTO7zP2c0pREZbw         0.440   0.417    0    -5.586     1   
459  7zjupUeHyd2j2FKlmSiZlj         0.454   0.790    6    -1.774     0   
520  72li0d2CgwciPLPADO0J0d         0.276   0.709    5    -6.496     0   
384  5hnyJvgoWiQUYZttV4wXy6         0.560   0.595    9    -4.530     1   
174  1UqhkbzB1kuFwt2iy4h29Q         0.534   0.721   11    -7.307     0   
474  1CI6YgwwNq3VfhgxPuEpE6         0.699   0.524   11    -7.522     1   
200  45iMRHmfLZu3yjdZ8P5e3t         0.672   0.642    5    -4.552     1   
341  0C4bx13JIMSWOAxd8TaoVq         0.496   0.880    1    -2.242     0   
404  57yI1u6t6cFXbjgjQcRHug         0.496   0.955    9    -3.607     0   
330  5bEkA2si3cnX6v66xOuxju         0.540   0.434    2    -8.937     1   
116  1BMLT5HPeIX2rxNvGAdUQD         0.747   0.371    9   -10.943     1   
33   1q1j4J77FDgq3fqnFCLSvi         0.400   0.320    0   -13.845     1   

     speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
104       0.0378      0.735000          0.002540    0.1110   0.4070  115.027   
521       0.0284      0.603000          0.000008    0.1010   0.3370   86.042   
526

In [177]:
bdf = bdf.append(df.loc[df['Liked?']==0], ignore_index=True)
bdf

track_uri  danceability  energy  key  loudness  mode  \
0   6mPpHS9ocuKjsehu8NgD7r         0.430   0.501    9    -8.864     1   
1   4urpFW2JCm486dHJmsQ8Ps         0.607   0.336    3   -11.854     1   
2   7c6TRb4cvgHWVhTJDfRqth         0.598   0.898   10    -5.155     1   
3   5AeoHJUx0PJXAzN425xryh         0.413   0.100    0   -16.349     1   
4   0M3adYbGtyRHACP86dey1H         0.426   0.888    0    -3.720     0   
..                     ...           ...     ...  ...       ...   ...   
95  1LJYn86ysceH708AIkw0VZ         0.975   0.482    9    -7.940     1   
96  7FERnkDYqD9D9DJdNjQlnf         0.761   0.246    0   -11.676     1   
97  4F0flTfcEU2ZcStYhJsaRy         0.862   0.717    9    -7.158     1   
98  15hJmqqEtASVXl6sM7i4UF         0.615   0.600   10    -5.620     1   
99  0lTqQTHcYb0RzGa5STjVrf         0.821   0.714    2    -5.977     1   

    speechiness  acousticness  instrumentalness  liveness  valence    tempo  \
0        0.0378      0.735000          0.002540    0.1110   0.4070  115.027   
1        0.0284      0.603000          0.000008    0.1010   0.3370   86.042   
2        0.1130      0.043300          0.000000    0.0976   0.4220   77.639   
3        0.0332      0.784000          0.381000    0.0819   0.1860  110.696   
4        0.0987      0.000455          0.000000    0.3060   0.3870  144.111   
..          ...           ...               ...       ...      ...      ...   
95       0.2460      0.007160          0.000000    0.0596   0.7370  125.953   
96       0.2450      0.132000          0.000000    0.0800   0.2260  130.012   
97       0.4600      0.166000          0.000000    0.1970   0.7750   77.620   
98       0.2700      0.107000          0.000002    0.2830   0.0661  130.027   
99       0.3740      0.109000          0.000000    0.2980   0.6650   81.393   

    duration_ms  time_signature  Liked?  
0        308373               4       1  
1        198219               4       1  
2        211933               4       1  
3        184907               3       1  
4        191520               4       1  
..          ...             ...     ...  
95       146598               4       0  
96       132989               4       0  
97       137806               4       0  
98       270671               4       0  
99       235583               4       0  

[100 rows x 15 columns]

And now that we have a our balanced sample, let's proceed with our data split

In [178]:
bX = bdf.drop(['track_uri', 'Liked?'], axis=1)
by = bdf['Liked?']
bX_train,bX_test,by_train,by_test=train_test_split(bX,by,test_size=0.25,random_state=0)
print(X_train)
print(y_train)

     danceability  energy  key  loudness  mode  speechiness  acousticness  \
89          0.575   0.765    9    -6.608     1       0.0345       0.00703   
285         0.484   0.793    2    -6.797     1       0.2110       0.00069   
310         0.625   0.715    7    -4.539     0       0.0942       0.04130   
367         0.529   0.934    0    -4.808     1       0.0602       0.00153   
46          0.410   0.233    0   -12.878     1       0.0286       0.83100   
..            ...     ...  ...       ...   ...          ...           ...   
277         0.674   0.204    5   -13.535     1       0.0384       0.88700   
9           0.544   0.224    7   -14.410     1       0.0351       0.78100   
359         0.397   0.903    2    -4.577     1       0.0457       0.00575   
192         0.708   0.413    7   -10.856     1       0.1010       0.78300   
559         0.499   0.160    7   -13.632     1       0.0362       0.89900   

     instrumentalness  liveness  valence    tempo  duration_ms  time_signat

And now our modelings to see if that performance is better.

In [179]:
blogmodel = LogisticRegression()
blogmodel.fit(bX_train,by_train)
bpredictions = blogmodel.predict(bX_test)
print(classification_report(by_test,bpredictions));

              precision    recall  f1-score   support

           0       0.47      0.67      0.55        12
           1       0.50      0.31      0.38        13

    accuracy                           0.48        25
   macro avg       0.49      0.49      0.47        25
weighted avg       0.49      0.48      0.46        25



In [180]:
bsvcmodel = SVC(kernel='linear')
bsvcmodel.fit(bX_train, by_train)
bpredictions = bsvcmodel.predict(bX_test)
print(classification_report(by_test,bpredictions))

              precision    recall  f1-score   support

           0       0.33      0.25      0.29        12
           1       0.44      0.54      0.48        13

    accuracy                           0.40        25
   macro avg       0.39      0.39      0.38        25
weighted avg       0.39      0.40      0.39        25



Now we have quite a few applications of classification models, which seems to be most accurate? The short answer is none of them are fantastic. However, the best of those shown depends on if FPs or FNs matter most. In our case neither really are all that great, we either listen to a song that we don't like or we miss songs we like that aren't recommended. I'd say as a result, the F-1 Score is the most important element of that evaluation. As such, we'd default to picking the modeling with the best F-1. This would be the SVC model with the weightings of 0.85 for our disliked and .15 for our liked songs to offset the imbalance between playlists.